FWCore/Framework/src/RunProcessingStatus.h

```diff
@@ -0,0 +1,123 @@
+// Class  :     RunProcessingStatus
+//
+/**\class edm::RunProcessingStatus
+
+ Description: Keep status information about one Run transition
+
+*/
+//
+// Original Author: W. David Dagenhart
+//          Created: 1 Oct 2021
+//
+
+#include "DataFormats/Provenance/interface/Timestamp.h"
+#include "FWCore/Concurrency/interface/LimitedTaskQueue.h"
+#include "FWCore/Concurrency/interface/WaitingTaskHolder.h"
+#include "FWCore/Concurrency/interface/WaitingTaskList.h"
+#include "FWCore/Framework/interface/InputSource.h"
+
+#include <atomic>
+#include <memory>
+#include <vector>
+
+namespace edm {
+
+  class EventSetupImpl;
+  class RunPrincipal;
+
+  class RunProcessingStatus {
+  public:
+    RunProcessingStatus(unsigned int iNStreams, WaitingTaskHolder const& holder);
+
+    WaitingTaskHolder& holderOfTaskInProcessRuns() { return holderOfTaskInProcessRuns_; }
+    void setHolderOfTaskInProcessRuns(WaitingTaskHolder const& holder) { holderOfTaskInProcessRuns_ = holder; }
+
+    void setResumer(LimitedTaskQueue::Resumer iResumer) { globalRunQueueResumer_ = std::move(iResumer); }
+    void resumeGlobalRunQueue() {
+      //free run for next usage
+      runPrincipal_.reset();
+      globalRunQueueResumer_.resume();
+    }
+
+    std::shared_ptr<RunPrincipal>& runPrincipal() { return runPrincipal_; }
+
+    void resetBeginResources();
+    void resetEndResources();
+
+    EventSetupImpl const& eventSetupImpl(unsigned subProcessIndex) const {
+      return *eventSetupImpls_.at(subProcessIndex);
+    }
+
+    EventSetupImpl const& eventSetupImplEndRun(unsigned subProcessIndex) const {
+      return *eventSetupImplsEndRun_.at(subProcessIndex);
+    }
+
+    std::vector<std::shared_ptr<const EventSetupImpl>>& eventSetupImpls() { return eventSetupImpls_; }
+    std::vector<std::shared_ptr<const EventSetupImpl>> const& eventSetupImpls() const { return eventSetupImpls_; }
+
+    WaitingTaskList& endIOVWaitingTasks() { return endIOVWaitingTasks_; }
+
+    std::vector<std::shared_ptr<const EventSetupImpl>>& eventSetupImplsEndRun() { return eventSetupImplsEndRun_; }
+    std::vector<std::shared_ptr<const EventSetupImpl>> const& eventSetupImplsEndRun() const {
+      return eventSetupImplsEndRun_;
+    }
+
+    WaitingTaskList& endIOVWaitingTasksEndRun() { return endIOVWaitingTasksEndRun_; }
+
+    void setGlobalEndRunHolder(WaitingTaskHolder holder) { globalEndRunHolder_ = std::move(holder); }
+    WaitingTaskHolder& globalEndRunHolder() { return globalEndRunHolder_; }
+
+    bool streamFinishedRun() { return 0 == (--nStreamsStillProcessingRun_); }
+
+    //These should only be called while in the InputSource's task queue
+    void updateLastTimestamp(edm::Timestamp const& iTime) {
+      if (iTime > endTime_) {
+        endTime_ = iTime;
+      }
+    }
+    edm::Timestamp const& lastTimestamp() const { return endTime_; }
+
+    void setEndTime();
+
+    void incrementStreamBeginRunTasksQueued() { ++streamBeginRunTasksQueued_; }
```

Comment:
In the current version of this PR, the function is safe because it is only called in one place (line of code) in a "for" loop that occurs only once for each RunProcessingStatus object. This is while the begin run stream tasks are being queued. And the data member is only being used after those tasks start to run and the queues are unpaused.

This is going to break when I replace PauseStreamQueuesSentry with a SerialQueue. So I will have to modify this somehow... I'm making a note of this and will modify both at the same time.